File: src/net/rizon/moo/tinc/node.scala
package net.rizon.moo.tinc

import scala.collection.mutable.LinkedList

object node
{
	private var nodes: LinkedList[node] = LinkedList()
	
	def getNode(serv: server, l: layer): node =
	{
		for (n <- nodes)
			if (n.getServer() == serv && n.getLayer == l)
				return n

		var n = new node(serv, l)
		nodes = nodes :+ n
		return n
	}
	
	def getNodes() = nodes
	
	def getNodesForLayer(l: layer): LinkedList[node] =
	{
		var nfl: LinkedList[node] = LinkedList()
		for (n <- nodes)
			if (n.getLayer() == l)
				nfl = nfl :+ n
		return nfl
	}
	
	def getNodeWithConfigName(l: layer, name: String): node =
	{
		for (n <- nodes)
			if (n.getLayer() == l && n.name.equals(name))
				return n
		return null
	}
}

class node (server: server, layer: layer)
{
	/* tinc.conf */
	var name = ""
	var pubAddress = ""
	/* connectTo in tinc.conf */
	var connectTo: LinkedList[String] = LinkedList()
	
	/* host config files on the machine */
	var hosts: LinkedList[host] = LinkedList()
	
	def getServer() = server
	def getLayer() = layer
	
	def getHost(n: String): host =
	{
		for (h <- hosts)
			if (h.name.equals(n))
				return h
		return null
	}
}

File: src/net/rizon/moo/tinc/node.java
package net.rizon.moo.tinc;

import java.util.ArrayList;
import java.util.LinkedList;

public class node
{
	private server server;
	private layer layer;
	
	/* tinc.conf */
	public String name;
	public String pubAddress;
	public LinkedList<String> connectTo = new LinkedList<String>();
	/* host config files on this machine */
	public LinkedList<host> hosts = new LinkedList<host>();
	
	public node(server serv, layer l)
	{
		this.server = serv;
		this.layer = l;
	}
	
	public server getServer()
	{
		return this.server;
	}
	
	public layer getLayer()
	{
		return this.layer;
	}
	
	public host getHost(String name)
	{
		for (host h : hosts)
			if (h.name.equals(name))
					return h;
		return null;
	}
	
	private static LinkedList<node> nodes = new LinkedList<node>();
	
	protected static node getNode(server serv, layer l)
	{
		for (node n : nodes)
			if (n.getServer() == serv && n.getLayer() == l)
				return n;
		
		node n = new node(serv, l);
		nodes.add(n);
		return n;
	}
	
	protected static node[] getNodes()
	{
		node[] n = new node[nodes.size()];
		nodes.toArray(n);
		return n;
	}
	
	public static node[] getNodesForLayer(layer l)
	{
		ArrayList<node> no = new ArrayList<node>();
		for (node n : nodes)
			if (n.getLayer() == l)
				no.add(n);
		node[] n = new node[no.size()];
		no.toArray(n);
		return n;
	}
	
	public static node getNodeWithConfigName(layer l, String name)
	{
		for (node n : nodes)
			if (n.getLayer() == l && n.name.equals(name))
				return n;
		return null;
	}
}